src/keyboard.c

```diff
@@ -231,6 +230,55 @@ handle_compositor_keybindings(struct wl_listener *listener,
 	return handled;
 }
 
+static int
+handle_keybind_repeat(void *data)
+{
+	struct keyboard *keyboard = data;
+	assert(keyboard->keybind_repeat);
+
+	/* synthesize event */
+	struct wlr_keyboard_key_event event = {
+		.keycode = keyboard->keybind_repeat_keycode,
+		.state = WL_KEYBOARD_KEY_STATE_PRESSED
+	};
+
+	handle_compositor_keybindings(keyboard, &event);
+	wl_event_source_timer_update(keyboard->keybind_repeat,
+		keyboard->keybind_repeat_rate);
+
+	return 0;
+}
+
+static void
+start_keybind_repeat(struct server *server, struct keyboard *keyboard,
+	struct wlr_keyboard_key_event *event)
+{
+	struct wlr_keyboard *wlr_keyboard = keyboard->wlr_keyboard;
+	assert(!keyboard->keybind_repeat);
+
+	if (wlr_keyboard->repeat_info.rate > 0
+			&& wlr_keyboard->repeat_info.delay > 0) {
+		struct wl_event_loop *event_loop =
+			wl_display_get_event_loop(server->wl_display);
```

Comment:
We have a pointer to the eventloop in `server->wl_event_loop` (see `server_init()` in `src/server.c`)

Reply:
Ah, didn't notice that, thanks